server/document.js

```diff
@@ -0,0 +1,118 @@
+import React, { Component, PropTypes } from 'react'
+import htmlescape from 'htmlescape'
+import { renderStatic } from 'glamor/server'
+import readPkgUp from 'read-pkg-up'
+
+const { pkg } = readPkgUp.sync({
+  cwd: __dirname,
+  normalize: false
+})
+
+export default class Document extends Component {
+  static getInitialProps ({ renderPage }) {
+    let head
+    const { html, css, ids } = renderStatic(() => {
+      const page = renderPage()
+      head = page.head
+      return page.html
+    })
+    const nextCSS = { css, ids }
+    return { html, head, nextCSS }
+  }
+
+  static childContextTypes = {
+    _documentProps: PropTypes.any
+  }
+
+  constructor (props) {
+    super(props)
+    const { __NEXT_DATA__, nextCSS } = props
+    if (nextCSS) __NEXT_DATA__.ids = nextCSS.ids
+  }
+
+  getChildContext () {
+    return { _documentProps: this.props }
+  }
+
+  render () {
+    return <html>
+      <Head />
+      <body>
+        <Main />
+        <NextScript />
+      </body>
+    </html>
+  }
+}
+
+export class Head extends Component {
+  static contextTypes = {
+    _documentProps: PropTypes.any
+  }
+
+  render () {
+    const { head, nextCSS } = this.context._documentProps
```

Comment:
`Head`, `Main` and `NextScript` receive `props` of `Document` via `context` so that you don't have to manually set `props` to these components. This should be convenient but can be too obscured ?